yarn-project/simulator/src/avm/journal/trace.ts

```diff
@@ -1,176 +1,278 @@
-    this.noteHashChecks.push(...incomingTrace.noteHashChecks);
-    this.newNoteHashes.push(...incomingTrace.newNoteHashes);
-    this.nullifierChecks.push(...incomingTrace.nullifierChecks);
-    this.newNullifiers.push(...incomingTrace.newNullifiers);
-    this.l1ToL2MessageChecks.push(...incomingTrace.l1ToL2MessageChecks);
-    this.newLogsHashes.push(...incomingTrace.newLogsHashes);
-    this.gotContractInstances.push(...incomingTrace.gotContractInstances);
+  public processFinishedChildTrace(incomingTrace: AvmSideEffectTrace) {
     // it is assumed that the incoming trace was initialized with this as parent, so accept counter
-    this.counter = incomingTrace.counter;
+    this.sideEffectCounter = incomingTrace.sideEffectCounter;
+    this.allUnencryptedLogs.push(...incomingTrace.allUnencryptedLogs);
+    // NOTE: eventually if the AVM circuit processes an entire enqueued call,
+    // this function will accept all of the child trace's side effects
+  }
+
+  /**
+   * Trace a nested call.
+   * Accept some results from a finished nested call's trace into this one.
+   */
+  public traceNestedCall(
+    /** The trace of the nested call. */
+    nestedCallTrace: AvmSideEffectTrace,
+    ...args: TraceNestedCallArgs
+  ) {
+    const result = nestedCallTrace.toExecutionResult(...args);
+    this.sideEffectCounter = result.endSideEffectCounter.toNumber();
+    // when a nested call returns, caller accepts its updated counter
+    this.allUnencryptedLogs.push(...result.allUnencryptedLogs.logs);
+    // NOTE: eventually if the AVM circuit processes an entire enqueued call,
+    // this function will accept all of the nested's side effects into this instance
+    this.nestedExecutions.push(result);
+
+    const gasUsed = new Gas(
+      result.startGasLeft.daGas - result.endGasLeft.daGas,
+      result.startGasLeft.l2Gas - result.endGasLeft.l2Gas,
+    );
+     this.avmCircuitHints.externalCalls.items.push(new AvmExternalCallHint(/*success=*/ new Fr(result.reverted ? 0 : 1), result.returnValues, gasUsed));
+
+  }
+
+  /**
+   * Convert this trace to a PublicExecutionResult for use externally to the simulator.
+   */
+  public toExecutionResult(
+    /** The execution that triggered this result. */
+    execution: PublicExecution,
+    /** How much gas was available for this public execution. */
+    startGasLeft: Gas,
+    /** How much gas was left after this public execution. */
+    endGasLeft: Gas,
+    /** Transaction fee set for this tx. */
+    transactionFee: Fr,
+    /** Bytecode used for this execution. */
+    bytecode: Buffer,
+    /** Calldata used for this execution. */
+    calldata: Fr[],
+    /** The return values of the function. */
+    returnValues: Fr[],
+    /** Whether the execution reverted. */
+    reverted: boolean,
+    /** The revert reason if the execution reverted. */
+    revertReason?: AvmRevertReason,
+  ): PublicExecutionResult {
+    return {
+      ...(this as FieldsOf<AvmSideEffectTrace>),
```

Comment:
this is kind of fine, but in some sense it feels it could be done outside of the simulator with some kind of `...journal.trace.flush()`

...of course... the external calls break everything!